Ab Test: Bug fixes and auto winner functionality

massudaw opened the pull request:
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### Steps to test this PR:
1. 
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

massudaw commented on app/bundles/EmailBundle/Entity/Email.php:
```diff
@@ -293,6 +296,14 @@ public static function loadMetadata(ORM\ClassMetadata $metadata)
 
         $builder->addPublishDates();
 
+        $builder->createField('autoRolloutDate', 'datetime')
```
Added fields, needs migrations .

massudaw commented on app/bundles/CampaignBundle/Entity/Event.php:
```diff
@@ -307,6 +317,7 @@ public static function loadApiMetadata(ApiMetadataDriver $metadata)
                      'triggerInterval',
                      'triggerIntervalUnit',
                      'triggerMode',
+                     'sampleSize',
```
added field needs migration

massudaw commented:
This pull request also depends on at least two code paths of the scalability plugin . 

https://github.com/MauldinEconomics/mautic-email-scalability-bundle/blob/master/Model/QueuedEmailModel.php#L681

https://github.com/MauldinEconomics/mautic-email-scalability-bundle/blob/master/Model/EventModelExtended.php#L798

maxlawton commented:
@massudaw ah, I see. So is that a hard dependency, or will this feature "work" without the plugin? By "work" I mean, as well as it would without the bugfixes in our plugin.

massudaw commented:
@maxlawton we would need to backport some of those changes to the core code in order the feature to work. The biggest problem of the code that does processing is that there were no modularization in it so we had to override a really big procedure to make our queue functionality work. 
If we commit this as an isolated function in core and just call it in the main processing code path of core and our plugins. Them is might be much easier to reduce the duplication and still commit the code to upstream.

maxlawton commented:
> If we commit this as an isolated function in core and just call it in the main processing code path of core and our plugins.

@massudaw Which part would be an isolated function, the processing additions from our plugin or the modified `convertVariant()` which calls the new `recordAbResult()` method?

massudaw commented:
@maxlawton   The processing additions from our plugin. We would introduce functions in the core code that does what we need, and call them in the plugins and core code.

maxlawton commented:
@massudaw Ok. Thanks for clarifying.

massudaw commented:
### Intro
The first thing about the changes for AB_TEST is that they are split into two main parts: 
 - Broadcast 
 - Campaigns
The Ab test has two main Groups. 
 - Sample 
 - Final Send
 
###  Broadcast
  For broadcast, the main place that we have the code is the QueuedEmailModel file. 
  This file is overriding some functions in the original QueuedEmailModel. There are changes which are related only to queueing. And some are related to AB_TEST.  
  - Split the method into two parts (Generator and Consumer) in order to implement Queueing. 
  - Improve pagination of the query, there is some overlapping with Mautic updates on pagination. They implemented something similar.
  - Split the code for processing broadcast into two stages when in AbTest.  Sampling and Final Send.
  - Modify queries to perform sampling. Note that we do sampling and pagination at the same time.
  - Check if the winner criteria are met and switch to final send.

  The general strategy for moving to the core is basically just splitting the code so it allows us to override the call for the consumer with the send to the queue. And leave the send email call in the core.
  We would introduce a function consume email that we can override with the logic. This is not so trivial, because there is also some resource management for the queue that needs to be done and we don't want to create the queue if there is no email to be sent. One possible solution i thought would be lazy load the queue in the first consumer call.


###  Campaigns 
  For campaigns we have features but it`s implemented quite differently. Because for campaigns we have multiple stages, Starting, Scheduled and Negative. For AB_TEST just Starting and scheduled is relevant. 
 -  All stages were split into two parts (Generator and Consumer) in order to implement Queueing. 
  - Improve pagination of the generator queries.  All generator queries were somewhat rewritten. But this does not affect AB_TEST.
  - Split the code for processing broadcast into two stages when in AbTest.  First part is done in Starting when is the AB_TEST event type.
  - The sampling is done in memory inside the Starting stage in the Generator loop.  For the Sample group we send directly the emails and for the remaining, we schedule the date that it will be sent using the scheduled events.
  - Check if the winner criteria are met then allow the final send. The check is done before the Scheduled loop and is where we pick which email to send.  Then all Final emails are sent as scheduled emails.

 For this change set, the path to moving to the core is a bit more convoluted. Because most of the changes were related to queueing. But should be possible to do the same strategy of creating a consumer function that we can override in our plugin. In this case, there are many more initializations, for campaigns and campaigns events that need to be handled in this splitting. And also the resource logic for the queue.  
 So if we can't move this splitting to the core. We might need to just move the functionality for AB_TEST into functions and use the functions to reimplement in the core code. Then we can share only the ab test functions in order to minimize the duplication.

massudaw commented:
@maxlawton  Check if this helps you.

maxlawton commented:
@massudaw So the majority of the AB Test functionality in our _Scalability Plugin_ is within the classes **QueuedEmailModel** (for _Broadcasts_) and **EventModelExtended** (for _Campaigns_), correct? It seems the **EventModelExtended** is much closer in structure to the core **EventModel**, whereas the queueing and AB Testing in the **QueuedEmailModel** seems significantly more involved and interconnected. 

Is there anything about the AB Test functionality in the _Broadcasts_ / **QueuedEmailModel** (or the _Campaigns_ for that matter) that will absolutely not work without the queueing process?

massudaw commented:
@maxlawton  Sorry I missed the notification. In both the queueing it's quite periferical. The only complication in QueuedEmailModel is that it has two ways to call it, console and web, and this makes  the code a bit more complicated.
All functionality about AB Test is independent of queueing. If you replace queueing by a function call. It should work normally.
